### PR TITLE
Add GitHub-style filter for columns (fix #37)

### LINKED ISSUE #37: [Feature request] Allow filter dates
how about allow search only forks updated the last 1 day / 1 week / 1 month / 1 year / custom?

### PR DESCRIPTION
**Pull request by Ethkuil**
There are 2 types of filter: `number` and `date`.
Tyey both support the following operators: `>`, `<`, `>=`, `<=`, `=`.

The format of date should be `yyyy-mm-dd`, like `2022-01-01`.

number: `stars`, `forks`, `ahead`, `behind`.
date: `pushed`

eg: `stars:>0 pushed:>2022-01-01`

---

Example screenshot: 
image

I have done tests on `payne911/PieMenu` and `Wilfred/difftastic`. It worked well. Certainly it would be better if more checks are taken.

---

It's a little troublesome because I'm more familiar with frameworks like Vue that supporting binding data with DOM, but here we have to update DOM ourselves and somewhere code about data and DOM are mixed.

I only did the necessary refactor to make it possible to add the feature of filtering.

---

And it would be much easier to implement "sort by columns"(#48 ) based on this PR. But I'm not so interested in that 🤔. I would be pleased if anyone would like to do that. Or I can try to solve it in spare time if needed.

### CODE REVIEW

**Comment by Ethkuil** [comment]
The UI may be not perfect because I'm not proficient in that. It's good to do some improvement if you think so.

**Comment by payne911** [comment]
First of all, thanks for this PR. 👍 

I mostly have three comments regarding it:
1. It seems like this change renders the "Ahead filter" of the "Query settings" dialog (opened with the clog icon) useless: consider simply removing it.
image
2. Is there a way to keep the current approach that adds a row even if the second batch of requests hasn't finished yet? I believe it's a better approach because it provides faster feedback to users, instead of letting them think the process is stalled or slow. Example image of what I'm talking about (this is taken during a query... it's obviously not the final result as a completed query should not have any of those "half-rows" left):
image
3. Your table rows have a border on the left-side: that would need to be removed. Here's a screenshot of your screenshot:
image

---

In a more general sense, some things may be worth discussing.

> It's a little troublesome because I'm more familiar with frameworks like Vue

Honestly, I've been wanting to rewrite the entire thing. Personally, I would probably opt for React because I've been wanting to learn to use it since it's the most used in the corporate world, but a rewrite in Vue could be just as interesting anyways. If you have the motivation to do it, go ahead (and I would definitely favor the usage of TypeScript).

The project was initially too small to justify using frontend frameworks, but now it has grown enough to justify the rewrite because:
* adding more features is becoming harder
* I believe the code is now ugly enough to generally demotivate people from submitting PRs
* it'd be great to add tests, and the current format is not conducive to that

> The UI may be not perfect [...] It's good to do some improvement if you think so.

I'm not very talented at it either, don't worry. We should probably discuss this a bit before you spend more time refining the current approach. Some ideas to consider:
* I think it would be good to at least use an icon to indicate the primary role of that new input field. Probably one of these 3 options (I searched "filter icon" on Google Image):
image
* Maybe the icon could remain on the right-side of the repo-input-field (although I think generally filters tend to be placed on the left side of query fields?), and the filter-input-field would be "toggleable" (i.e. expand/retract) on-click of that icon.
* Or another approach could be closer to this mock-up, where that "header row" would always be displayed, and which includes the column sorting as well (clickable triangles, the red one is the active one). Each of those `>0` (and others) are input fields, and I think it's a good idea to have pre-filled values because it makes it more obvious how to use them in terms of syntax.
image

**Comment by Ethkuil** [comment]
Thanks for your detailed comments.

> 1. It seems like this change renders the "Ahead filter" of the "Query settings" dialog (opened with the clog icon) useless: consider simply removing it.

I think so too, but it's backwards incompatible, so that I didn't remove it before just waiting for your opinion.

> 2. Is there a way to keep the current approach that adds a row even if the second batch of requests hasn't finished yet? I believe it's a better approach because it provides faster feedback to users, instead of letting them think the process is stalled or slow.

I think the rising call count is enough to tell users everything goes well. What's more, it's a little tricky to do that: we need to add a fork into data array before we confirm it's indeed useful, so that we have to remove it from the array. The cost is too expensive for me.

> 3. Your table rows have a border on the left-side: that would need to be removed.

There is a misunderstanding. Maybe it's just because I mouse over the inappropriate position.

> 4. I haven't had time to verify this, but make sure the filtering would not prevent rows from registering themselves (what I mean is that if filtering is applied, and then removed, the entries that were filtered out should reappear).

Certainly I have tested it.

---

I largely agree with you. I hope that this project could be refactored with a modern framework, whether utilizing Vue or React. At the same time a corresponding UI component library would be much helpful. With these, we can simply choose "option 3" after all it's a common feature that we don't need to implement it again.

If you think it's OK, maybe I can refactor it in the spare time, using Vue3, Element UI Plus as UI library (because I'm more familiar with them lol) and TypeScript (I love it!). I cannot guarantee a relatively short time frame to begin and finish, but I think it'll be shorter than half a year.

And by the way we can also use tools based on React if you think it's better, because I also have a plan to learn it.

**Comment by payne911** [comment]
> but it's backwards incompatible

It's possible to make a retrocompatible change: after removing the option, you can nonetheless reuse the saved value and use that as the default.
Here's the relevant code pointer: https://github.com/useful-forks/useful-forks.github.io/blob/master/website/src/settings.js#L22-L36
I think it'd be fine to have only the `ahead` value as saveable considering it's the most relevant and probably-recurrent preference.
In any case, that change can be its own PR, so you don't really have to bother with it just yet if you don't want.

> I think the rising call count is enough to tell users everything goes well.

Fair enough. (Although I liked how active and dynamic it looked.)

----

I've tested your branch. There are a few bugs that would need to be ironed out:
1. If you input ridiculous filters like `ahead:<0` or `ahead:>9999999` before running the query, it will result in an empty table, and the tool will thus display the message reserved for when there are no forks to display (img 1). But then if you remove the filter and there were actually forks, the message remains (img 2).
image
image
Once the scan is done, this message should be updated dynamically based on the filtered results. The same goes for the "Export as CSV" button.
2. Submitting a search filters while the progress bar is displayed (i.e. when the scan stalls, e.g. with https://useful-forks.github.io/?repo=libgdx/libgdx) will remove that message. It's imperative that it remains.
3. The corners of the input fields are broken. Both of these should be rounded:
image

Comments:
* Consider not allowing people to input filters until the request ends (either via success, or API rate limit).
* Am I understanding correctly that for every fork row added to the table, you recreate the entire table? I noticed "slow-downs" for queries that result in big tables (e.g. https://useful-forks.github.io/?repo=libgdx/libgdx) when you move the mouse around on the rows (which highlights them on hover). The lag isn't there when the scan stalls (and thus isn't periodically recreating the table).
* There should now be two CSV buttons: one to download the subset of filtered results, and another one to download the full list.
* I'm not sure why, but the fact that the filter input field has no button would make me think that it applies automatically. It took me a split second to realize I needed to press 'Enter' to submit.
* The syntax bothers me a bit. Why require an extra "`:`" character to specify values?

---

Regarding the rewrite:

> With these, we can simply choose "option 3" after all it's a common feature that we don't need to implement it again.

Can you point me toward some specific documentation displaying what this common feature would look like?

> we can also use tools based on React if you think it's better, because I also have a plan to learn it.

Then this looks like the perfect opportunity to learn it. ;)

**Comment by Ethkuil** [comment]
Something easy will be replied first as following:

> ... in any case, that change can be its own PR, so you don't really have to bother with it just yet if you don't want. For now, the preference can remain.

In fact I prefer to simply delete it, too.

> 3. The corners of the input fields are broken. Both of these should be rounded:

> 4. For smaller screens, there is an overlap. That's another good argument to move the filter box elsewhere.

I have made a simple adjustment to fix them. Do you think the following is satisfying?

image

---

The bug 1 and 2 are both about "messages". I will try to write a commit named "Fix wrong message display after filtering".

---

> Consider not allowing people to input filters until the request ends (either via success, or API rate limit).

I think the current is better because you may hope that useless forks won't appear in the all process. And we can achieve that.

> Am I understanding correctly that for every fork row added to the table, you recreate the entire table? I noticed "slow-downs" for queries that result in big tables (e.g. [useful-forks.github.io/?repo=libgdx/libgdx](https://useful-forks.github.io/?repo=libgdx/libgdx)) when you move the mouse around on the rows (which highlights them on hover). The lag isn't there when the scan stalls (and thus isn't periodically recreating the table).

It recreate the entire table iff the table data is updated (iff the request for a list of forks succeed) or the filter is updated.

I'm not sure why it slowed down when moving the mouse around on the rows. I'm sorry but could you confirm that so it is or just a coincidence?

> There should now be two CSV buttons: one to download the subset of filtered results, and another one to download the full list.

I never use this feature so ignored it before... The button now can download the subset of filtered results. I'm not sure whether the other button is necessary or not, but certainly you can add one if need.

> I'm not sure why, but the fact that the filter input field has no button would make me think that it applies automatically. It took me a split second to realize I needed to press 'Enter' to submit.

You're right. I have amended it.

> The filtering syntax bothers me a bit. Why require an extra ":" character to specify values?

The reason lies within the title: Add **GitHub-style** filter ...
 
> Pointer for adding an icon in your input field: [bulma.io/documentation/form/general/#with-icons](https://bulma.io/documentation/form/general/#with-icons)

image

To supplement a prompt in placeholder could be simpler and is also universally used.

---

> Can you point me toward some specific documentation displaying what this common feature would look like?

[Element Plus - Table - filter](https://element-plus.org/en-US/component/table.html#filter)
[Element Plus - Table - custom header](https://element-plus.org/en-US/component/table.html#table-with-custom-header)

> > we can also use tools based on React if you think it's better, because I also have a plan to learn it.
> Then this looks like the perfect opportunity to learn it. ;)

Certainly it's all right, if you can accept that I need to spend much more time to finish the job because of learning lol.

**Comment by payne911** [comment]
> I prefer to simply delete it

Sounds good. Although let's discuss that in another PR after we are with this one.

---

A few things:
1. The `is-dark` class makes the filter field seem as important as the input field, and it also prevents a mouse-hover effect. But I tried removing it and it looks kind of too pale... I wish there was an in-between. What do you think?
image
2. The filter field feels a bit weird being so large when in fact most people will barely input more than 2 keywords in it.
3. The filter field should be positioned in between "Export CSV" and the table. This would essentially lead to fixing the problem mentioned just above (about the sizing) because then it will be more obvious that the sizing should match the table's width.
4. Due to the immediate filtering, using `keyword:<` will filter everything out. Can we avoid that by only applying filters once at least a digit is placed after the symbols?
5. Rename the CSV button to "Export filtered table as CSV", but only when a filter is applied.
6. There should be an indicator of how many rows are being hidden by the filters. It could be the same kind of little badge as the one used to display the amount of calls.
7. There should probably be an "X" button at the right end of the filter field to clear the field.
8. Still missing a filter icon on the left end of the filter field.

Points 3 to 6 are probably the most critical ones.

> To supplement a prompt in placeholder could be simpler and is also universally used.

I'm not sure what you mean.

>  I will try to write a commit named "Fix wrong message display after filtering".

Sounds good to me.

> I'm sorry but could you confirm that so it is or just a coincidence?

I can't reproduce the problem anymore, and it was pretty minute anyways. Let's ignore this for now.

---

While your linked examples look great, I think it'd be wiser to use React. It's much more widely adopted within the industry, and so would probably increase the likelihood of future contributions.

**Comment by Ethkuil** [comment]
I'm sorry for late response because I'm just too busy this week. Tired and nearly crazy.

"Fix wrong message display after filtering" has been committed, which I promised but hadn't finished before.

---

> 1. The filter field should be positioned in between "Export CSV" and the table. (And make sure it only appears when there is at least 1 row that exists.)

I agree that it should be positioned under the row "Queried repository". And "between 'Export CSV' and the table" is also not bad for me.

(But I almost wanted to cry when I saw how "useful_forks_inject" was built. 😥I haven't changed them because I'm not familiar with JQuery.)

But in my opinion, **it's reasonable for users to edit the filter field before the searching finishes** and at this time the filter should also work.

> 2. Due to the immediate filtering, using keyword:< will filter everything out. Can we avoid that by only applying filters once at least a digit is placed after the symbols?

Done.

> 3. Rename the CSV button to "Export filtered table as CSV", but only when a filter is applied.

My idea: "Export the following table as CSV". The simple is better. Do not increase complexity unnecessarily.

> 4. There should be an indicator of how many rows are being hidden by the filters. It could be the same kind of little badge as the one used to display the amount of calls.

Would you mind proving an example that this feature helps?

---

>> To supplement a prompt in placeholder could be simpler and is also universally used.
>
> I'm not sure what you mean.

I mean, a suitable placeholder is enough and maybe we don't need a filter icon. Certainly it's also OK if added, but I won't do that myself.

---

Certainly I'm convinced that React is powerful enough. After all, developers won't choose a framework if it's so weak. I will try to learn React and use it in `useful-forks` in my spare time.

**Comment by payne911** [comment]
> I'm sorry for late response

This is all voluntary. No worries! :)

Note: before testing your PR, I merged my latest changes in. They do not affect your logic.

---

> [Can we avoid that by only applying filters once at least a digit is placed after the symbols?] Done.

You introduced a bug. Here's a sequence of events to demonstrate:

image
image
image
image

Only the currently-being-built token should not be treated. The other ones are fine.

> My idea: "Export the following table as CSV".

Actually a good idea. Although I'd prefer shorter. How about "Export table below as CSV" ?

> Would you mind providing an example that this feature helps?

You can ignore the request. I'll add it myself if I feel motivated enough, I guess.

> but I won't do that myself.

Same as above.

> I will try to learn React

Great news 👍

**Comment by Ethkuil** [comment]
> You introduced a bug. Here's a sequence of events to demonstrate...

I make some changes to achieve as below:
```js
/**
 * 1. Empty filter means no filter.
 * 2. Filter string is a list of conditions separated by spaces.
 * 3. If a condition is invalid, it is ignored, and the rest of the conditions are applied.
 */
```

> How about "Export table below as CSV" ?

It's actually better. I replaced the string for both the button and the note in settings.

**Comment by payne911** [comment]
I like the rules you opted for. 👍 

Two last things:
1. The position of the filter field
2. Consider adding aliases for each columns (for ahead, behind, date, name, stars, forks: ``a``, ``b``, ``d``, ``n``, ``s``, ``f``)

**Comment by Ethkuil** [comment]
> 1. The position of the filter field

I don't know how to edit
```js
$('#useful_forks_inject').append(
    $('<div>', {id: UF_ID_WRAPPER}).append(
        $('<div>', {id: UF_ID_HEADER}),
        $('<div>', {id: UF_ID_MSG}).html(landingPageTrigger()),
        $('<div>', {id: UF_ID_DATA}).append(
            $('<table>', {id: UF_ID_TABLE}).append(
                $('<tbody>')
            )
        )
    )
);
```

> 2. Consider adding aliases for each columns (for ahead, behind, date, name, stars, forks: a, b, d, n, s, f)

I feel it a bit "no so useful". After all, `useful-forks` is not a tool used daily, like `Git`. I will prefer to wait for the novel type of table that allows filter for each column directly.

But it's easy to implement so I can do that if you think it's useful indeed.

> 3. Is there a link to the "GitHub-style" filters documentation we could point users to?

It's here: [about-searching-on-github](https://docs.github.com/en/search-github/getting-started-with-searching-on-github/about-searching-on-github)

However, I think the syntax is simple enough for users to get.

**Comment by Ethkuil** [comment]
> The position of the filter field

Sorry I decide to give up. The filter field should be positioned under the msg about searching, but if so then it'll be under the Introducing msg. Different types of msgs occupy the same area.

I consider making do is enough for now.

> If it's an easy grab, it's worth it.

Done.

**Comment by payne911** [comment]
This might help you achieve the goal of not always displaying the filter field:

```
const FILTER_DIV_ID = "uf_filter_field_div";
const FILTER_INPUT_ID = "uf_filter_field";
const INPUT_FIELD = $('<input>', {
                       class: 'input is-dark',
                       type: 'text',
                       id: FILTER_INPUT_ID,
                       placeholder: 'Type your filter here. eg: ahead:>=1 behind:<5'
                    })
const FILTER_DIV = $('<div>', {id: FILTER_DIV_ID, class: "mt-2"}).append(INPUT_FIELD);
```

---

Also, I found another messaging bug: when we input a repo which doesn't exist, the returned message is `UF_MSG_EMPTY_FILTER` instead of `UF_MSG_ERROR`.

image

**Comment by Ethkuil** [comment]
> Also, I found another messaging bug: when we input a repo which doesn't exist, the returned message is UF_MSG_EMPTY_FILTER instead of UF_MSG_ERROR.

It has been fixed now and I have checked the code relate to msg management to avoid the similar bugs.

**Comment by Ethkuil** [comment]
> This might help you achieve the goal of not always displaying the filter field

Thank you. However, it's not my goal. I still think it's reasonable for users to input a filter even if the searching results haven't come.

**Comment by payne911** [comment]
> I still think it's reasonable for users to input a filter even if the searching results haven't come.

This may be fair, but I disagree with the idea that this extra input field should appear at all times.

When a new user lands on the home page, it should only see the repository field because that's all they should care about at that point. Another field is just confusing.

Again: I'm fine with showing the filter field during the scan, but not before it begins. Nor should it be displayed when there are no useful forks.

Essentially, it should be displayed once this condition is fulfilled: https://github.com/useful-forks/useful-forks.github.io/blob/master/website/src/queries-logic.js#L316

File: website/index.html [hunk 1]
```diff
@@ -230,7 +230,10 @@
               Find useful forks
             </button>
           </p>
-        </div>
+          </div>
```
Comment (payne911) [inline website/index.html:233]: The alignment of those `div`s is wrong.
Plus, as discussed, the filter field shouldn't appear when there isn't even a repository provided yet.

File: website/src/queries-init.js [hunk 2]
```diff
@@ -21,7 +22,6 @@ const UF_MSG_API_RATE     = "<b>GitHub API rate-limits exceeded.</b> Consider pr
 // list of messages which should not be cleared when the request ends
 const UF_PRESERVED_MSGS = [
     UF_MSG_NO_FORKS,
-    UF_MSG_EMPTY_FILTER,
```
Comment (payne911) [inline website/src/queries-init.js:24]: Why did you need to take this one out?

File: website/src/queries-logic.js [hunk 3]
```diff
@@ -135,9 +137,19 @@ function allRequestsAreDone() {
...
   if (allRequestsAreDone()) {
-    clearNonErrorMsg();
-    sortTable();
+    manageMsgs();
     enableQueryFields();
+  }
+}
+
+function manageMsgs() {
+  clearNonErrorMsg()
```
Comment (payne911) [inline website/src/queries-logic.js:146]:
```suggestion
  clearNonErrorMsg();
```

File: website/src/queries-logic.js [hunk 4]
```diff
@@ -164,41 +176,26 @@ function send(requestPromise, successFn, failureFn) {
...
+/** Updates table data, then calls function to update the table. */
+function update_table_data(responseData, user, repo, parentDefaultBranch, is_useful_fork) {
+  if (isEmpty(responseData)) {
     return;
+  }
 
-  if (!RATE_LIMIT_EXCEEDED) // because some times gets called after some other msgs are displayed
+  if (!RATE_LIMIT_EXCEEDED) {// because some times gets called after some other msgs are displayed
     clearNonErrorMsg();
+  }
```
Comment (payne911) [inline website/src/queries-logic.js:187]:
```suggestion
  if (!RATE_LIMIT_EXCEEDED) { // because some times gets called after some other msgs are displayed
    clearNonErrorMsg();
  }
```

File: website/src/queries-logic.js [hunk 5]
```diff
@@ -207,27 +204,22 @@ function add_fork_elements(forkdata_array, user, repo, parentDefaultBranch) {
...
-          setMsg(UF_MSG_EMPTY_FILTER);
+      if (responseData.total_commits > AHEAD_COMMITS_FILTER) {
+        datum['ahead_by'] = responseData.ahead_by;
+        datum['ahead_url'] = responseData.html_url;
+        datum['behind_by'] = responseData.behind_by;
+        datum['behind_url'] = getBehindUrl(responseData.html_url);
+        datum['pushed_at'] = getOnlyDate(currFork.pushed_at);
+        TABLE_DATA.push(datum);
+
+        if (typeof is_useful_fork === 'function') {
```
Comment (payne911) [inline website/src/queries-logic.js:215]: Excuse my ignorance: why do we need to check this is a function? (i.e. when would this not be a function?)

File: website/src/queries-logic.js [hunk 6]
```diff
@@ -237,6 +229,119 @@ function add_fork_elements(forkdata_array, user, repo, parentDefaultBranch) {
   }
 }
 
+function update_filter() {
+  let is_useful_fork = getFilterFunction()
+  if (typeof is_useful_fork === 'function') {
+    update_table(TABLE_DATA.filter(is_useful_fork));
+  } else {
+    update_table(TABLE_DATA);
+  }
```
Comment (payne911) [inline website/src/queries-logic.js:238]: Seems like this is repeated twice. Maybe could be extracted as a method?

File: website/src/queries-logic.js [hunk 7]
```diff
@@ -164,41 +176,26 @@ function send(requestPromise, successFn, failureFn) {
...
-          $('<td>').html(UF_TABLE_SEPARATOR + getForkCol(num_forks)).attr("value", num_forks)
-      )
-  );
-  return NEW_ROW;
-}
-
-/** Add bold to the date text if the date is earlier than the queried repo. */
-function compareDates(date, html) {
-  return REPO_DATE <= new Date(date) ? `<strong>${html}</strong>` : html;
-}
```
Comment (payne911) [inline website/src/queries-logic.js:183]: Would you mind bringing this piece of code back above that method so that Git doesn't think it was touched by this PR?

**Review by payne911 — CHANGES_REQUESTED** [review]
Generally speaking, this seems to make the code cleaner. I appreciate the changes.

Could you consider extracting that table-related logic and filter-related logic into their own class or files?

That way, `queries-logic.js` would remain mostly about the logic handling the GitHub API requests. It's a bit of a particular file because `npm run dev` needs to be called to test any change to this file, so it's nice if we can get more code out of it (while also having the great benefit of better encapsulation).

File: website/index.html (hunk 1 above)
Comment (Ethkuil) [reply website/index.html:233]: > The alignment of those divs is wrong.

After formatting, you can see that in fact the alignment of those elements is wrong originally.

> Plus, as discussed, the filter field shouldn't appear when there isn't even a repository provided yet.

You will be satisfied after new commits.

**Comment by Ethkuil** [comment]
I disabled the automatic formatting function because I found that it will affect almost all the original code. And I'm glad to fix that now because you also noticed that.

File: website/src/queries-logic.js (hunk 5 above)
Comment (Ethkuil) [reply website/src/queries-logic.js:220]: When the function is called without parameter of `is_useful_fork` (so it's undefined). And I think to check whether it's a function is a little better to check whether it's undefined because we only need function.

You will see `undefined is not a function` error in console if we remove the check.

File: website/src/queries-logic.js (hunk 6 above)
Comment (Ethkuil) [reply website/src/queries-logic.js:243]: I used to try and my idea about the name is `update_table_trying_use_filter`. But I finally abort since it doesn't "deserve".

Certainly, it's also reasonable to extract that as a method because its meaning is also clear. If you stick.

File: website/src/queries-init.js (hunk 2 above)
Comment (Ethkuil) [reply website/src/queries-init.js:24]: Why did you think it "error"?

```js
// list of messages which should not be cleared when the request ends
const UF_PRESERVED_MSGS = [
  UF_MSG_NO_FORKS,
  UF_MSG_ERROR,
  UF_MSG_API_RATE
];
```

```js
function clearNonErrorMsg() {
  const msg = JQ_ID_MSG.html();
  if (!UF_PRESERVED_MSGS.includes(msg))
    clearMsg();
}
```

Firstly let's ignore the gap between `error` and `preserved`.

When the filter updates, certainly we need to clear some outdated msgs and only preserve. And the msg `empty` could be outdated because the table may switch between empty and non-empty as I edit the filter field.

**Comment by Ethkuil** [comment]
> Again: I'm fine with showing the filter field during the scan, but not before it begins. Nor should it be displayed when there are no useful forks.
> 
> Essentially, at the earliest it could be displayed once this condition is fulfilled: [master/website/src/queries-logic.js#L316](https://github.com/useful-forks/useful-forks.github.io/blob/master/website/src/queries-logic.js?rgh-link-date=2023-03-06T04%3A14%3A19Z#L316)
> 
> ... but it'd be even better if it was only displayed when this one is called at least twice (i.e. at least two useful forks are listed): [master/website/src/queries-logic.js#L221](https://github.com/useful-forks/useful-forks.github.io/blob/master/website/src/queries-logic.js?rgh-link-date=2023-03-06T04%3A14%3A19Z#L221)
> 
> And of course, when any of those 3 messages are displayed, it would make no sense to display that field either: [master/website/src/queries-init.js#L7-L9](https://github.com/useful-forks/useful-forks.github.io/blob/master/website/src/queries-init.js?rgh-link-date=2023-03-06T04%3A14%3A19Z#L7-L9)

(By the way, " UF_MSG_EMPTY_FILTER" should not be considered. If you filter out all forks and now you want to delete some conditions, I consider you won't want to see the filter field has gone.)

I hope the the current version can satisfy you. And the solution is:

1. At the beginning, hidden.
2. After a new searching begins, hidden.
3. When there have been more than 1 useful forks, shown.

I have tested it and feel satisfied. How about you?

File: website/src/queries-init.js (hunk 2 above)
Comment (payne911) [reply website/src/queries-init.js:24]: Ah, I must've forgotten to adapt the method's name at some point.
You are right that it shouldn't mention "error".

File: website/src/queries-logic.js (hunk 6 above)
Comment (payne911) [reply website/src/queries-logic.js:244]: `update_table_trying_use_filter` sounds good enough to me. Feel free to update with it.

**Comment by payne911** [comment]
> the codebase itself hasn't been something clean

Absolutely agreed. The codebase is messy, among other things because I began the project by forking someone else's project. That led to two different styles mixing together, and initially the changes were so minimal I didn't bother with applying a linter. Then the project grew, and I never really paid much attention to that aspect because it was just a sideline hobby written in a language I pretty much never use.

Thus, it's mostly random styles, and since a rewrite to React is indeed something I'm considering, I don't plan on spending time on fixing the current issues related to that.

To conclude: please revert [11334a5](https://github.com/useful-forks/useful-forks.github.io/pull/58/commits/11334a50cfe8131943303bca33b8e631ba3a8a54) : I'd rather not have the `git blame` affected globally like that (plus, it makes it harder to review your PR).

File: website/index.html [hunk 8]
```diff
@@ -230,7 +230,11 @@
...
+          <div class="control" id="filterContainer" hidden>
+            <input class="input is-dark" type="text" id="filter" name="filter"
+              placeholder="Type your filter here. eg: ahead:>=1 behind:<5" />
+          </div>
```
Comment (payne911) [inline website/index.html:237]:
```suggestion
        </div>
        <div class="control" id="filterContainer" hidden>
          <input class="input is-dark" type="text" id="filter" name="filter"
            placeholder="Type your filter here. eg: ahead:>=1 behind:<5" />
        </div>
```

**Review by payne911 — CHANGES_REQUESTED** [review]

**Review by payne911 — CHANGES_REQUESTED** [review]
There is one bug left before I merge this:

image

The progress bar disappears when the filters are edited while it's displayed. This is critical.

**Comment by payne911** [comment]
And a follow-up for myself:
- [ ] Add "filter icon" (see https://github.com/useful-forks/useful-forks.github.io/pull/58#issuecomment-1426857970)
- [ ] Remove the "ahead_filter" from the clog settings
- [ ] Immediate visual feedback based on filter input (`is-dark` class maybe only when there is valid input? and red color when one of the inputs is broken?)
- [ ] Allow alternate syntax (`:` feels extremely superfluous and annoying to add)
- [ ] Add indicator of how many rows are being hidden by the filters (same kind of little badge as the one used to display the amount of calls)
- [ ] Provide an "X" button at the right end of the filter field to clear the field

File: website/index.html [hunk 9]
```diff
@@ -231,6 +231,10 @@
             </button>
           </p>
         </div>
+        <div class="control" id="filterContainer" hidden>
+          <input class="input is-dark" type="text" id="filter" name="filter"
+            placeholder="Type your filter here. eg: ahead:>=1 behind:<5" />
```
Comment (payne911) [inline website/index.html:236]:
```suggestion
          <input class="input is-dark" type="text" id="filter" name="filter"
                 title='GitHub style syntax, with some aliases like "a" for "ahead", "d" for the date, etc. (see https://docs.github.com/en/search-github/getting-started-with-searching-on-github/about-searching-on-github)'
                 placeholder="Type your filter here. eg: ahead:>=1 behind:<5" />
```

**Review by payne911 — CHANGES_REQUESTED** [review]

**Comment by Ethkuil** [comment]
> The progress bar disappears when the filters are edited while it's displayed. This is critical.

I didn't even know its existence... I felt a little impotent lol.

I checked how the progress bar is added and removed. Then I thought "maybe the progress bar shouldn't be contained in messages". Based on the idea I fixed the bug. I have used reference in VSCode to confirm I have edited everywhere need to be edited and nowhere else.

**Comment by payne911** [comment]
Sorry for the miscommunication. The message associated with the progress bar is important as well because it explains why the user has to wait.

Currently, only the progress bar remains.
image

**Comment by payne911** [comment]
You seem to have fixed it. Well done! :)

I recorded another bug: https://imgur.com/a/2XJRch5

It seems to happen only when the filter is altered while requests are being issued.

**Comment by payne911** [comment]
I think when I removed the filter just before it showed the progress bar, it decided not to apply the filter. I had to enter something, and then remove the content again.

image

File: website/src/queries-logic.js [hunk 10]
```diff
@@ -164,41 +181,40 @@ function send(requestPromise, successFn, failureFn) {
...
+}
+
+/** Updates table data, then calls function to update the table. */
+function update_table_data(responseData, user, repo, parentDefaultBranch) {
+  if (isEmpty(responseData)) {
     return;
+  }
 
-  if (!RATE_LIMIT_EXCEEDED) // because some times gets called after some other msgs are displayed
+  if (!RATE_LIMIT_EXCEEDED) {// because some times gets called after some other msgs are displayed
```
Comment (payne911) [inline website/src/queries-logic.js:203]:
```suggestion
  if (!RATE_LIMIT_EXCEEDED) { // because some times gets called after some other msgs are displayed
```

**Review by payne911 — CHANGES_REQUESTED** [review]

File: website/src/queries-logic.js [hunk 11]
```diff
@@ -237,6 +254,121 @@ function add_fork_elements(forkdata_array, user, repo, parentDefaultBranch) {
...
+    }
+    const [_, operator, value] = requirement.split(/([<>=]+)/);
+    if (!operator || !value) {
+      continue; // invalid condition
+    }
+    if (attribute in mapTable) {
+      attribute = mapTable[attribute];
+    }
+    conditionObj[attribute] = { operator, value };
+  }
```
Comment (payne911) [inline website/src/queries-logic.js:341]: What are your thoughts on this? Why prefer the GitHub syntax when we can support something simpler?
```suggestion
  for (const condition of conditionStrList) {
    const [attribute, operator, value] = condition.match(/(\w+)(<=|>=|[<=>])([0-9]+)/)?.slice(1);
    if (!attribute || !operator || !value) {
      continue; // invalid condition
    }
    if (attribute in mapTable) {
      attribute = mapTable[attribute];
    }
    conditionObj[attribute] = { operator, value };
  }
```

Comment (Ethkuil) [reply website/src/queries-logic.js:341]: All right. GitHub and Google chose such syntax may be because users usually search about string, and so the filters should be distinguished from that. Considering we don't need to search about any strings, a simpler syntax can be better indeed.

**Comment by Ethkuil** [comment]
> I found another bug: using Ctrl + X or backspace to clear the entire text in the filter prevents it from being re-applied.

It's undoubtedly a foolish bug. I have fixed it.

**Comment by payne911** [comment]
I think your filtering code doesn't treat dates properly. Could you explain to me a bit how it handles them?
I thought one needed to convert with `Date()` to allow proper comparison handling in JS.

**Comment by payne911** [comment]
Please review my latest commits.
If you are fine with the current state of the PR, I'll go ahead and merge it.

**Comment by Ethkuil** [comment]
> Could you explain to me a how your filtering/comparison code handles dates?

It's just "as is". Before this PR, this project has provided date info as string, like `2020-11-03`. And string comparison works well as date comparison here.

For example, we consider `d>2020-01-01`. The first 5 chars, `2020-`, same. And the 6th char, `1` and `0` where `'1' > '0'`. Therefore, we get `2020-11-03 > 2020-01-01`.

Certainly, if we input `d>2020`, then `2020-11-03` will be kept too (equivalent to `d>=2020-01-01`). And `d>2021` will filter out it. I think this is enough.

**Comment by Ethkuil** [comment]
> Please review my latest commits.
> If you are fine with the current state of the PR, I'll go ahead and merge it.

The current version can not deal with `d>2021-04`.

I would like to change the regex to `/([a-z]+)(<=|>=|[<=>])(\S+)/`. It's more flexible, whereas it will lead to something else becomes different: if you type condition like `d>2021p`, the previous version will ignore it whereas this version will filter out all results because nothing can meet this requirement.

**Comment by payne911** [comment]
Interesting consideration. I think we could get to an even better regex, although definitely more complex.

```
[0-9]{4}(?:(?<!-|-[0-9])-[0-9]{0,2}){0,2}
```

* The non-capturing group (`(?:`) is used to not screw up with the value extraction
* The negative look-behind (`(?<!`) is used to not accept a match on `2020--` and `2020-1-`

**Comment by payne911** [comment]
# Fun fact
Here is the final regex in all of its (monstrous) glory:
```
([a-z]+)(<=|>=|[<=>])([0-9]{4}(?:(?<!-|-[0-9])-[0-9]{0,2}){0,2}|[0-9]+)
```

**Comment by payne911** [comment]
@Ethkuil Congratulation: your PR made it through the review process. 🥳

Thanks again for your contribution! ❤️

**Comment by Ethkuil** [comment]
I want to touch base with you regarding our discussion about rewriting the repository with React.js. After considering my current workload and priorities, I have come to the realization that I won't be able to contribute to this project as much as I had initially hoped.

While there is a possibility that I may spend some time on this, the likelihood is low so that you can ignore it for now. (I'll keep you updated if I do decide to work on this.)

I apologize for any inconvenience or disappointment this may cause. Thank you for your understanding, and I wish you all the best with the project.

Best regards.

**Comment by payne911** [comment]
> I apologize

No worries. Like I said: I planned on doing it myself anyways since it's a good learning opportunity. And it's not like you signed any contract.

I'm very glad to have collaborated with you. May your talent help other projects move forward. :)
